src/Ocelot/DownstreamRouteFinder/Finder/DownstreamRouteFinder.cs

```diff
@@ -47,10 +53,33 @@ public Response<DownstreamRouteHolder> Get(string upstreamUrlPath, string upstre
             return new ErrorResponse<DownstreamRouteHolder>(new UnableToFindDownstreamRouteError(upstreamUrlPath, httpMethod));
         }
 
-        private static bool RouteIsApplicableToThisRequest(Route route, string httpMethod, string upstreamHost)
+        private static bool RouteIsApplicableToThisRequest(Route route, string httpMethod, string upstreamHost, IHeaderDictionary requestHeaders)
         {
-            return (route.UpstreamHttpMethod.Count == 0 || route.UpstreamHttpMethod.Select(x => x.Method.ToLower()).Contains(httpMethod.ToLower())) &&
-                   (string.IsNullOrEmpty(route.UpstreamHost) || route.UpstreamHost == upstreamHost);
+            return (route.UpstreamHttpMethod.Count == 0 || RouteHasHttpMethod(route, httpMethod)) &&
+                   (string.IsNullOrEmpty(route.UpstreamHost) || route.UpstreamHost == upstreamHost) &&
+                   (route.UpstreamHeaderRoutingOptions == null || !route.UpstreamHeaderRoutingOptions.Enabled() || RouteHasRequiredUpstreamHeaders(route, requestHeaders));
+        }
+
+        private static bool RouteHasHttpMethod(Route route, string httpMethod)
+        {
+            httpMethod = httpMethod.ToLower();
+            return route.UpstreamHttpMethod.Select(x => x.Method.ToLower()).Contains(httpMethod);
+        }
+        
+        private static bool RouteHasRequiredUpstreamHeaders(Route route, IHeaderDictionary requestHeaders)
+        {
+            bool result = false;
+            switch (route.UpstreamHeaderRoutingOptions.Mode)
+            {
+                case UpstreamHeaderRoutingTriggerMode.Any:
+                    result = route.UpstreamHeaderRoutingOptions.Headers.HasAnyOf(requestHeaders);
+                    break;
+                case UpstreamHeaderRoutingTriggerMode.All:
+                    result = route.UpstreamHeaderRoutingOptions.Headers.HasAllOf(requestHeaders);
+                    break;
+            }
+
+            return result;

```

Comment:
The **switch**-sentence makes sense if there will be 3+ options or 2 + default.
We have 2 options only in the enumeration. So we can use conditional flow:
```suggestion
            var options = route.UpstreamHeaderRoutingOptions;
            return options.Mode == UpstreamHeaderRoutingTriggerMode.Any
                ? options.Headers.HasAnyOf(requestHeaders)
                : options.Headers.HasAllOf(requestHeaders);
```

Reply:
✅ Fixed by [lines 68-70](files#diff-9c51794162197f4d2f9ad8f40e3155d9a4228b32e09b8cc76d5e10f77e6682cbR68-R70)